File: relay/export-contacts.js
// JXA script: export Contacts.app entries to ~/claudeclaw/relay/contacts.json
// Run via: osascript -l JavaScript export-contacts.js

ObjC.import('Foundation');

function normalizePhone(s) {
  return String(s || '').replace(/[^\d+]/g, '');
}

const Contacts = Application('Contacts');
const people = Contacts.people();

const out = [];
for (let i = 0; i < people.length; i++) {
  const p = people[i];
  let first = ''; let last = ''; let org = ''; let nick = '';
  try { first = p.firstName() || ''; } catch (e) {}
  try { last = p.lastName() || ''; } catch (e) {}
  try { org = p.organization() || ''; } catch (e) {}
  try { nick = p.nickname() || ''; } catch (e) {}

  let name = '';
  if (first && last) name = first + ' ' + last;
  else if (first) name = first;
  else if (last) name = last;
  else if (org) name = org;
  else if (nick) name = nick;

  if (!name) continue;

  const phones = [];
  try {
    const ps = p.phones();
    for (let j = 0; j < ps.length; j++) {
      let v = '';
      try { v = ps[j].value() || ''; } catch (e) {}
      const n = normalizePhone(v);
      if (n) phones.push(n);
    }
  } catch (e) {}

  const emails = [];
  try {
    const es = p.emails();
    for (let j = 0; j < es.length; j++) {
      let v = '';
      try { v = es[j].value() || ''; } catch (e) {}
      if (v) emails.push(String(v).toLowerCase());
    }
  } catch (e) {}

  if (phones.length === 0 && emails.length === 0) continue;
  out.push({ name, org: org || null, phones, emails });
}

const fm = $.NSFileManager.defaultManager;
const json = JSON.stringify(out, null, 0);
const data = $(json).dataUsingEncoding($.NSUTF8StringEncoding);
const path = $('~/claudeclaw/relay/contacts.json').stringByExpandingTildeInPath.js;
const ok = data.writeToFileAtomically(path, true);

JSON.stringify({ contacts: out.length, written: ok, path });

File: relay/export-contacts.swift
#!/usr/bin/env swift
// Bulk-export Contacts.app entries to JSON.
// Uses the native Contacts framework — ~100x faster than osascript/JXA
// for 1000+ contact databases.
//
// Usage: swift export-contacts.swift > contacts.json

import Contacts
import Foundation

let store = CNContactStore()

// Request access synchronously (will silently succeed if already granted)
let semaphore = DispatchSemaphore(value: 0)
var hasAccess = false
store.requestAccess(for: .contacts) { granted, _ in
    hasAccess = granted
    semaphore.signal()
}
semaphore.wait()

if !hasAccess {
    FileHandle.standardError.write("ERROR: Contacts access denied. Grant Privacy & Security → Contacts to the calling binary.\n".data(using: .utf8)!)
    exit(2)
}

let keys: [CNKeyDescriptor] = [
    CNContactGivenNameKey as CNKeyDescriptor,
    CNContactFamilyNameKey as CNKeyDescriptor,
    CNContactOrganizationNameKey as CNKeyDescriptor,
    CNContactNicknameKey as CNKeyDescriptor,
    CNContactPhoneNumbersKey as CNKeyDescriptor,
    CNContactEmailAddressesKey as CNKeyDescriptor,
]

func normalizePhone(_ s: String) -> String {
    return s.filter { ("0"..."9").contains($0) || $0 == "+" }
}

var rows: [[String: Any]] = []

let request = CNContactFetchRequest(keysToFetch: keys)
do {
    try store.enumerateContacts(with: request) { contact, _ in
        let first = contact.givenName
        let last = contact.familyName
        let org = contact.organizationName
        let nick = contact.nickname

        var name = ""
        if !first.isEmpty && !last.isEmpty { name = "\(first) \(last)" }
        else if !first.isEmpty { name = first }
        else if !last.isEmpty { name = last }
        else if !org.isEmpty { name = org }
        else if !nick.isEmpty { name = nick }

        if name.isEmpty { return }

        var phones: [String] = []
        for p in contact.phoneNumbers {
            let n = normalizePhone(p.value.stringValue)
            if !n.isEmpty { phones.append(n) }
        }

        var emails: [String] = []
        for e in contact.emailAddresses {
            let v = (e.value as String).lowercased()
            if !v.isEmpty { emails.append(v) }
        }

        if phones.isEmpty && emails.isEmpty { return }

        var row: [String: Any] = [
            "name": name,
            "phones": phones,
            "emails": emails,
        ]
        if !org.isEmpty { row["org"] = org }
        rows.append(row)
    }
} catch {
    FileHandle.standardError.write("ERROR: \(error.localizedDescription)\n".data(using: .utf8)!)
    exit(3)
}

let data = try JSONSerialization.data(withJSONObject: rows, options: [])
if let s = String(data: data, encoding: .utf8) {
    print(s)
}
FileHandle.standardError.write("exported \(rows.count) contacts\n".data(using: .utf8)!)
